Topics/Tech_Stacks/Pythonic-Iteration-Patterns.md

```diff
@@ -0,0 +1,238 @@
+`zip` also supports an optional `strict` keyword argument to raise an error
+if the given iterables do not have the same number of items to iterate over.
+Also see [`zip_longest` from the `itertools` module](
+https://docs.python.org/3/library/itertools.html#itertools.zip_longest).
+
+
+### `reversed`
+```python
+s = 'defghi'
+
+# Instead of:
+for i in range(len(s) - 1, -1, -1):
+    print(s[i])
+
+# Or (this creates a copy of the string, which can be inefficient):
+for char in s[::-1]:
+    print(char)
+
+# Write:
+for char in reversed(s):
+    print(char)  # prints 'i', then 'h', then 'g', then 'f', then 'e', then 'd'
+```
+
+
+### `pairwise`
+```python
+s = 'monoid'
+
+# Instead of:
+for i in range(len(s) - 1):
+    print(f'{s[i]}|{s[i + 1]}')
+
+# Write:
+import itertools
+
+for char_a, char_b in itertools.pairwise(s):
+    print(f'{char_a}|{char_b}')  # prints 'm|o', then 'o|n', then 'n|o', then 'o|i', then 'i|d'
+```
+In general, [the `itertools` module](https://docs.python.org/3/library/itertools.html)
+provides quite a few functions that can be helpful when trying to iterate over things.
+
+
+
+## Already done `for` you
+
+Python doesn't just have tools to make it nicer to write for loops and comprehensions --
+it also has a variety of functions that make it so that you don't _need_ to write those.
+
+
+### Constructors: comprehension shortcuts
+
+First off, instead of writing `[x for x in iterable]` or `{x for x in iterable}`,
+you can simply write `list(iterable)` and `set(iterable)`. Similar functionality also exists for `tuple` and `dict`.
+
+
+### Grabbag of convenient functions
+
+Python comes with functions to compute the sum, product, maximum, or minimum of any iterable:
+`sum`, `prod` from the `math` module, `max`, and `min`.
+If you want to check if a criterion is satisfied for every item in an iterable, use the `all` function;
+for at least one item in an iterable, the `any` function.
+
+
+### `str.join`
+
+I've seen some variation of the following code from more than one person in relation to my team's project:
+```python
+result = ''
+
+for s in ['a', 'monad', 'is', 'a', 'monoid']:
+    result += s + ' | '
+
+# remove the trailing ' | '
+# a better way to accomplish this would be the str.removesuffix method
+result = result[:-3]
+
+# result == 'a | monad | is | a | monoid'
+```
+Instead of writing that, simply write:
+```python
+result = ' | '.join(['a', 'monad', 'is', 'a', 'monoid'])
+```
+
+
+### Removing the square brackets
+
+When using all of these functions, instead of giving a list comprehension as an argument, consider using a
+[generator comprehension](https://stackoverflow.com/questions/364802/how-does-a-generator-comprehension-works)
+instead -- instead of generating an entire list up front,
+items are computed on demand, making things more memory-efficient.
+
+For example:
+```python
+lst = [3, 1, 4, 1, 5, 9]
+
+# Instead of:
+sum([x * 2 for x in lst])
+
+# Simply write:
+sum(x * 2 for x in lst)
+```

```

Comment:
The link above is good (enables the reader to have a good understanding of the generator), however, it'd be better to add more details in this example.
Compared to the examples in the link, the example here seems to be confusing.